src/stdlib_32_bit_nmhashes.fypp

```diff
@@ -0,0 +1,801 @@
+
+contains
+
+    pure function nmh_readle32( p ) result( v )
+        integer(int32) :: v
+        integer(int8), intent(in) :: p(1:4)
+
+        if ( little_endian ) then
+            v = transfer( p(1:4), 0_int32 )
+        else
+            v = transfer( [ p(4), p(3), p(2), p(1) ], 0_int32 )
+        end if
+
+    end function nmh_readle32
+
+    pure function nmh_readle16( p ) result( v )
+        integer(int16) :: v
+        integer(int8), intent(in) :: p(1:2)
+
+        if ( little_endian ) then
+            v = transfer( p(1:2), 0_int16 )
+        else
+            v = transfer( [ p(2), p(1) ], 0_int16 )
+        end if
+
+    end function nmh_readle16
+
+    pure function nmhash32_0to8( x, seed ) result( vx32 )
+        integer(int32), intent(in) :: x
+        integer(int32), intent(in) :: seed
+        integer(int32) :: vx32
+        ! base mixer: [-6 -12 776bf593 -19 11 3fb39c65 -15 -9 e9139917 -11 16]
+        ! = 0.027071104091278835
+        integer(int32), parameter :: m1 = int(z'776BF593', int32)
+        integer(int32), parameter :: m2 = int(z'3FB39C65', int32)
+        integer(int32), parameter :: m3 = int(z'E9139917', int32)
+
+        integer(int16) :: vx16(0:1)
+        integer(int16), parameter :: &
+            m116(0:1) = transfer( m1, 0_int16, 2 ), &
+            m216(0:1) = transfer( m2, 0_int16, 2 ), &
+            m316(0:1) = transfer( m3, 0_int16, 2 )
+
+        vx32 = x
+        vx32 = ieor( vx32, ieor( ishft( vx32, -12 ), ishft( vx32, -6 ) ) )
+        vx16 = transfer( vx32, 0_int16, 2 )
+        vx16 = vx16 * m116
+        vx32 = transfer( vx16, 0_int32 )
+        vx32 = ieor( vx32, ieor( ishft( vx32, 11 ), ishft( vx32, -19 ) ) )
+        vx16 = transfer( vx32, 0_int16, 2 )
+        vx16 = vx16 * m216
+        vx32 = transfer( vx16, 0_int32 )
+        vx32 = ieor( vx32, seed )
+        vx32 = ieor( vx32, ieor( ishft( vx32, -15 ), ishft( vx32, -9 ) ) )
+        vx16 = transfer( vx32, 0_int16, 2 )
+        vx16 = vx16 * m316
+        vx32 = transfer( vx16, 0_int32 )
+        vx32 = ieor( vx32, ieor( ishft(vx32, 16), ishft(vx32, -11) ) )
+
+    end function nmhash32_0to8
+
+    pure function nmhash32_9to255( p, seed, full_avalanche ) result( hash )
+        integer(int8), intent(in)  :: p(0:)
+        integer(int32), intent(in) :: seed
+        logical, intent(in)        :: full_avalanche
+        integer(int32) :: hash
+
+        integer(int32) :: xu32(0:3), yu32(0:3)
+        integer(int16) :: xu16(0:1)
+        integer(int16), parameter :: &
+            nmh_m1_16(0:1) = transfer( nmh_m1, 0_int16, 2 ),  &
+            nmh_m2_16(0:1) = transfer( nmh_m2, 0_int16, 2 ),  &
+            nmh_m3_16(0:1) = transfer( nmh_m3, 0_int16, 2 )
+        integer(int32) :: s1
+        integer(int64) :: length
+        integer(int32) :: length32(0:1)
+        integer(int64) :: i, j, r
+
+        ! base mixer: [f0d9649b  5 -13 29a7935d -9 11 55d35831 -20 -10 ] =
+        ! 0.93495901789135362
```

Comment:
I don't understand this comment

Reply:
It is from the original code. I think it implies that the code is equivalent to multiplying by the floating point number and rounding (down?).